### PR TITLE
Fuzz 30087

### PR DESCRIPTION
**Pull request by jkbonfield**
Fix an overflow bug in unstripe function.

Also move some common functions into utils.h.

### CODE REVIEW

File: htscodecs/utils.h [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+/*
+ * Order 1 histogram construction.  4-way unrolled to avoid cache collisions.
+ */
+static inline
+void hist1_4(unsigned char *in, unsigned int in_size,
+	     uint32_t F0[256][256], uint32_t *T0) {
+    int T1[256+MAGIC] = {0}, T2[256+MAGIC] = {0}, T3[256+MAGIC] = {0};
+    int i;
```
Comment (daviesrob) [inline htscodecs/utils.h:142]: There's another `i` declared later when merging `F0` and `F1`.  Possibly this declaration should be moved to just above the loop merging the `T` arrays, where it's actually used?

File: htscodecs/utils.h [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+    unsigned char l = 0, c;
+    unsigned char *in_end = in + in_size;
+
+    unsigned char cc[5] = {0};
+    if (in_size > 500000) {
+	int F1[256][259] = {{0}};
```
Comment (daviesrob) [inline htscodecs/utils.h:149]: uint32_t to match `F0`.

**Review by daviesrob — COMMENTED** [review]
I think this will work as-is, but it would be nice to tidy up some of the signed / unsigned inconsistencies to make the code a little easier to understand and check.

File: htscodecs/utils.h [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+/*
+ * Data transpose by N.  Common to rANS4x16 and arith_dynamic decoders.
+ *
+ * Tuned for specific common cases of N.
+ */
+static inline void unstripe(unsigned char *out, unsigned char *outN,
+			    unsigned int ulen, unsigned int N,
+			    unsigned int idxN[256]) {
+    int j = 0, k;
```
[the remark this answers is not in the document]
Comment (jkbonfield) [reply htscodecs/utils.h:44]: Maybe, but it's slower so I'll keep it as is I think.

File: htscodecs/utils.h [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+	default:
+	    // General case, around 25% slower overall decode
+	    while (j < ulen-N) {
+		for (k = 0; k < N; k++)
+		    out[j++] = outN[idxN[k]++];
+	    }
+	    break;
+	}
+    }
+    while (j < ulen) {
```
[the remark this answers is not in the document]
Comment (jkbonfield) [reply htscodecs/utils.h:71]: Agreed

File: htscodecs/utils.h [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+/*
+ * Order 0 histogram construction.  8-way unrolled to avoid cache collisions.
+ */
+static inline
+void hist8(unsigned char *in, unsigned int in_size, uint32_t F0[256]) {
+    uint32_t F1[256+MAGIC] = {0}, F2[256+MAGIC] = {0}, F3[256+MAGIC] = {0};
+    uint32_t F4[256+MAGIC] = {0}, F5[256+MAGIC] = {0}, F6[256+MAGIC] = {0};
+    uint32_t F7[256+MAGIC] = {0};
+
+    int i, i8 = in_size & ~7;
```
[the remark this answers is not in the document]
Comment (jkbonfield) [reply htscodecs/utils.h:88]: They don't seem to have any performance hit, so will change.

File: htscodecs/utils.h (hunk 1 above)
Comment (jkbonfield) [reply htscodecs/utils.h:142]: Agreed

**Comment by jkbonfield** [comment]
Applied changes bar the unstripe sign.  Making it unsigned penalises both gcc and clang.

For example:

```
# signed int
./tests/rans4x16pr -o136 -t /var/tmp/q4
171.8 MB/s enc, 1175.1 MB/s dec	 15000000 bytes -> 963622 bytes
174.1 MB/s enc, 1176.3 MB/s dec	 15000000 bytes -> 963622 bytes
174.3 MB/s enc, 1176.4 MB/s dec	 15000000 bytes -> 963622 bytes
173.7 MB/s enc, 1176.1 MB/s dec	 15000000 bytes -> 963622 bytes
```

```
#unsigned int
./tests/rans4x16pr -o136 -t /var/tmp/q4
172.6 MB/s enc, 1090.7 MB/s dec	 15000000 bytes -> 963622 bytes
174.3 MB/s enc, 1091.5 MB/s dec	 15000000 bytes -> 963622 bytes
172.8 MB/s enc, 1066.9 MB/s dec	 15000000 bytes -> 963622 bytes
173.3 MB/s enc, 1091.1 MB/s dec	 15000000 bytes -> 963622 bytes
174.3 MB/s enc, 1091.6 MB/s dec	 15000000 bytes -> 963622 bytes
```

Given it's just for neatness and tidiness, I propose we keep the faster code.

**Comment by jkbonfield** [comment]
It's perhaps also worth pointing out that htslib ought to be building with -O3, as htscodecs does.  The Makefile here using -O3 and the code has been optimised assuming that.  However replacing the Makefile here with htslib's own undoes that and drops back to O2 instead.  I don't really understand why htslib deliberately picks a lower optimisation, but definitely this code benefits from O3.  It's not huge though - a few percent.

**Comment by daviesrob** [comment]
An interesting result.  In my testing it went considerably faster if you use `size_t` instead of `int`.  You might want to change to that in a future update.
```
# signed int
Testing 15 blocks
107.2 MB/s enc, 670.3 MB/s dec	 15000000 bytes -> 963622 bytes
135.6 MB/s enc, 1006.6 MB/s dec	 15000000 bytes -> 963622 bytes
145.4 MB/s enc, 1021.3 MB/s dec	 15000000 bytes -> 963622 bytes
149.3 MB/s enc, 1040.9 MB/s dec	 15000000 bytes -> 963622 bytes
143.3 MB/s enc, 1019.2 MB/s dec	 15000000 bytes -> 963622 bytes
141.6 MB/s enc, 1079.1 MB/s dec	 15000000 bytes -> 963622 bytes
155.3 MB/s enc, 1083.5 MB/s dec	 15000000 bytes -> 963622 bytes
155.5 MB/s enc, 1082.1 MB/s dec	 15000000 bytes -> 963622 bytes
155.6 MB/s enc, 1067.9 MB/s dec	 15000000 bytes -> 963622 bytes
155.6 MB/s enc, 1075.4 MB/s dec	 15000000 bytes -> 963622 bytes

# size_t
Testing 15 blocks
153.7 MB/s enc, 1478.7 MB/s dec	 15000000 bytes -> 963622 bytes
155.2 MB/s enc, 1502.9 MB/s dec	 15000000 bytes -> 963622 bytes
155.4 MB/s enc, 1501.5 MB/s dec	 15000000 bytes -> 963622 bytes
155.5 MB/s enc, 1490.2 MB/s dec	 15000000 bytes -> 963622 bytes
155.0 MB/s enc, 1491.5 MB/s dec	 15000000 bytes -> 963622 bytes
151.9 MB/s enc, 1499.7 MB/s dec	 15000000 bytes -> 963622 bytes
151.9 MB/s enc, 1483.5 MB/s dec	 15000000 bytes -> 963622 bytes
153.7 MB/s enc, 1497.5 MB/s dec	 15000000 bytes -> 963622 bytes
153.5 MB/s enc, 1452.4 MB/s dec	 15000000 bytes -> 963622 bytes
154.7 MB/s enc, 1492.7 MB/s dec	 15000000 bytes -> 963622 bytes
```

**Comment by daviesrob** [comment]
I've rebased with the to-merge parts merged into the appropriate commits.

**Comment by jkbonfield** [comment]
Wow that's a huge speed increase.  I'll explore.  Great spot!

**Pull request merged**